File: Hosting/Digitalocean/Digitalocean.1h.rb
#!/usr/bin/ruby
# coding: utf-8
#
# <bitbar.title>Digitalocean</bitbar.title>
# <bitbar.version>v0.1</bitbar.version>
# <bitbar.author>Alejandro Torres</bitbar.author>
# <bitbar.desc>List your Digitalocean droplets.</bitbar.desc>
# <bitbar.image>http://i.imgur.com/GV9FXrE.png</bitbar.image>
# <bitbar.dependencies>ruby</bitbar.dependencies>
# <bitbar.abouturl>https://github.com/fulgorek</bitbar.abouturl>

require 'json'
require 'net/https'

module DigitalOcean
  class Client
    DIGITALOCEAN_API = 'https://api.digitalocean.com'

    attr_reader :access_token

    def initialize(options = {})
      @access_token = options[:access_token] || no_access_token
      @get = options[:get] || 'no_method'
      self.send(@get) if defined? @get
      render_info
    end


    def droplets
      show_toolbar_icon
      fetch("#{DIGITALOCEAN_API}/v2/droplets")
      data = JSON.parse(@response.body)
      no_droplets_handle if data['droplets'].empty?
      # process_droplets(data)
      fake
      create_droplet
    end

    private

    def render_info
      puts '---'
      puts 'Web console | href=https://cloud.digitalocean.com'
    end

    def create_droplet
      puts '---'
      puts 'Create Droplet | href=https://cloud.digitalocean.com/droplets/new?size=512mb'
    end

    def show_toolbar_icon
      puts '|image=iVBORw0KGgoAAAANSUhEUgAAAA4AAAAOCAQAAAC1QeVaAAAA4klEQVQYGQXBMSuEAQAG4Oe77066H0CJ0tmkSAaFQbFjUUaT36AMSpKSbNhkNhmslCx3ZTAohYiUsjql83qeAqUOFiwaUvPh0qkvpQ5V9LkSERHxaQkl9HgREU/utUXECnAhomVaXZeGIxE/GsyKuFUFwK6IAw5EzKFboVBD3bu4r+jHqxb+lPhV+tZEgz4jBpUACiXOxG8BYMyEHxe+UPNg0DMDNkxjT8QkWBdxzLzYxo6IUWyK+DNc6DFuxpRrN+qWzegFa44osC+2wKOIb6uoVFV0nLjTRJc3becOvSh1/gGbP1qNI14GZAAAAABJRU5ErkJggg=='
      puts '---'
    end

    def process_droplets(data)
      data['droplets'].map do |d|
        name = d['name']
        region = d['region']['slug']
        ipv4 = d['networks']['v4'][0]['ip_address']
        color = d['status'] == 'active' ? '#222222' : '#FF0000'
        text = "[#{ipv4}] #{name}"
        spaces = (17 - text.split(' ').first.length + 2)
        puts text.gsub(' ', ' ' * spaces) + " |color=#{color}"
      end
    end

    def open_config_file
      `/usr/bin/open -t #{File.dirname(__FILE__)}`
    end

    def no_access_token
      show_toolbar_icon
      puts 'No access token! | color=red'
      puts "Open config file | bash=\"#{open_config_file}\" terminal=false refresh=false\n"
      exit
    end

    def no_method
      puts 'Please set method'
    end

    def handle_errors
      puts 'Error!'
      exit
    end

    def no_droplets_handle
      puts 'No Droplets!'
      create_droplet
      render_info
      exit
    end

    def fetch(uri)
      uri = URI.parse(uri)
      https = Net::HTTP.new(uri.host, uri.port)
      https.use_ssl = true
      req = Net::HTTP::Get.new(uri.path, initheader = connection_options)
      @response = https.request(req)
      handle_errors if @response.code != '200'
    end

    def connection_options
      {
        'Content-Type' => 'application/json',
        'Authorization' => "Bearer #{access_token}"
      }
    end
  end
end

DigitalOcean::Client.new({
  :access_token => '',
  :get => 'droplets'
})

File: Dev/Hosting/DigitalOcean/digitalocean.1h.js
#!/usr/bin/env /usr/local/bin/node

/* jshint esversion: 6 */

/*
 * <bitbar.title>DigitalOcean</bitbar.title>
 * <bitbar.version>v1.0</bitbar.version>
 * <bitbar.author>Hasit Mistry</bitbar.author>
 * <bitbar.author.github>hasit</bitbar.author.github>
 * <bitbar.image>http://i.imgur.com/1YTWFSn.png</bitbar.image>
 * <bitbar.desc>View status of DigitalOcean Droplets</bitbar.desc>
 * <bitbar.dependencies>node</bitbar.dependencies>
 */

var https = require('https');

var hostname = 'api.digitalocean.com';

// Add your Digital Ocean token. 
// If you have different DO teams you can add multiple tokens: ['DO_TOKEN_1', 'DO_TOKEN_2', ...]
var DOTokens = ['DO_TOKEN'];

var menuBarIcon = '💧';
var startMenu = '---';

console.log(menuBarIcon);
console.log(startMenu);

for (var i = 0; i < DOTokens.length; i++) {
    if (DOTokens.length == 1 && DOTokens[0] === 'DO_TOKEN') {
        console.log('DO token is missing!');
    } else {
        tokenIndex = i;
        getDroplets();
    }
}

function getDroplets() {
    performRequest('/v2/droplets', 'GET', function(data) {
        var droplets = data.droplets;

        for (var i = 0; i < droplets.length; i++) {
            var status = 'iVBORw0KGgoAAAANSUhEUgAAAAwAAAAMCAMAAABhq6zVAAAAXVBMVEUAAADAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvAOSvGu33lAAAAHnRSTlMAAwYHCQoQEh0fNVBRX2N5fJG+wMfK3O3v8fP3+/37J2BkAAAATklEQVQIHQXBhQHCAADAsA53d1j+P5OkqoaqqmbHr/G8rGr1BuOmmjwB47L2AM51A/CrFwBDFwDf2gI41fAAfObV4gFe66qG3fV9P0yrPxEfCr3MVhkLAAAAAElFTkSuQmCC';
            var date = new Date(droplets[i].created_at);
            var humanDate = date.getFullYear() + '/' + date.getMonth() + '/' + date.getDate() + ' ' + date.getHours() + ':' + date.getMinutes();

            if (droplets[i].status === 'active') {
                status = 'iVBORw0KGgoAAAANSUhEUgAAAAwAAAAMCAMAAABhq6zVAAAAXVBMVEUAAAAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmAnrmCynqnaAAAAHnRSTlMAAwYHCQoQEh0fNVBRX2N5fJG+wMfK3O3v8fP3+/37J2BkAAAATklEQVQIHQXBhQHCAADAsA53d1j+P5OkqoaqqmbHr/G8rGr1BuOmmjwB47L2AM51A/CrFwBDFwDf2gI41fAAfObV4gFe66qG3fV9P0yrPxEfCr3MVhkLAAAAAElFTkSuQmCC';
            }

            printBitBarLine(droplets[i].name, ['color=black', 'image=' + status]);

            // Open in browser and console
            printBitBarLine('--Open in browser', ['href=https://cloud.digitalocean.com/droplets/' + droplets[i].id]);
            printBitBarLine('--Connect to console', ['href=https://cloud.digitalocean.com/droplets/' + droplets[i].id + '/console?no_layout=true']);

            // Droplet creation date and time
            printBitBarLine('-----');
            printBitBarLine('--Created');
            printBitBarLine('--' + humanDate, ['color=black']);

            // Droplet IPv4 address
            printBitBarLine('-----');
            printBitBarLine('--IP Address');
            printBitBarLine('--' + droplets[i].networks.v4[0].ip_address, ['color=black']);

            // Droplet region
            printBitBarLine('-----');
            printBitBarLine('--Region');
            printBitBarLine('--' + droplets[i].region.name, ['color=black']);

            // Droplet details - image, memory, storage, vcpus
            printBitBarLine('-----');
            printBitBarLine('--Details');
            printBitBarLine('--Image: ' + droplets[i].image.distribution + droplets[i].image.name, ['color=black']);
            printBitBarLine('--Memory: ' + droplets[i].memory + 'MB', ['color=black']);
            printBitBarLine('--Storage: ' + droplets[i].disk + 'GB', ['color=black']);
            printBitBarLine('--CPUs: ' + droplets[i].vcpus, ['color=black']);
        }
    });
}

function printBitBarLine(title, args) {
    var lineArgs = [];

    if (title !== '-----') {
        for (var i in args) {
            lineArgs.push(args[i]);
        }

        if (lineArgs.length !== 0) {
            title = title + '|';
        }
    }

    console.log(title + lineArgs.join(' '));
}

function performRequest(endpoint, method, success) {
    var headers = {};

    headers = {
        'Authorization': 'Bearer ' + DOTokens[tokenIndex],
        'Content-Type': 'application/json'
    };

    var options = {
        host: hostname,
        path: endpoint,
        method: method,
        headers: headers
    };

    var req = https.request(options, function(res) {
        res.setEncoding('utf-8');

        var responseString = '';

        res.on('data', function(data) {
            responseString += data;
        });

        res.on('end', function() {
            var responseObject = JSON.parse(responseString);
            success(responseObject);
        });
    });

    req.end();

    req.on('error', function(err) {
        console.log(err);
    });
}
